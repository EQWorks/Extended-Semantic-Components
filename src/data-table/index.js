import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Table, Pagination, Button, Form } from 'semantic-ui-react'
import numeral from 'numeral'

import DataTableColumn, {
  propTypes as columnProps,
  defaultProps as columnDefaultProps,
} from './data-table-column'
import customSort from '../utils/sort'
import defaultSearch from '../utils/search'

const colPropKeys = Object.keys(columnProps)

const childrenColumnCheck = (props, propName, componentName) => {
  if (props.children && props.columns) {
    return new Error(`Only one or none of 'children' or 'columns' is allowed in '${componentName}'`)
  }
}

const propTypes = {
  data: PropTypes.array.isRequired,
  children: childrenColumnCheck,
  columns: childrenColumnCheck,
  defaultSortKey: PropTypes.string,
  /** one of ['descending', 'ascending'] */
  defaultSortDir: PropTypes.string,
  downloadName: PropTypes.string,
  download: PropTypes.bool,
  perPage: PropTypes.number,
  onRowClick: PropTypes.func,
  isRowActive: PropTypes.func,
  emptySearchMsg: PropTypes.string,
  noColumnsMsg: PropTypes.string,
  downloadPicked: PropTypes.bool,
  search: PropTypes.func,
}

const defaultProps = {
  defaultSortKey: '',
  defaultSortDir: 'descending',
  downloadName: 'Table',
  download: true,
  perPage: 9,
  onRowClick: null,
  isRowActive: null,
  emptySearchMsg: 'Couldn\'t find anything :(',
  noColumnsMsg: 'No columns selected',
  downloadPicked: false,
  search: null,
}


class DataTable extends Component {
  constructor(props) {
    super(props)

    const {
      defaultSortKey: sortColumn,
      defaultSortDir: sortDirection,
    } = props

    const picked = this.pickables()
    this.state = {
      sortColumn,
      sortDirection,
      activePage: 1,
      searchInput: '',
      picked,
    }
  }

  downloadReport = () => {
    const { data, downloadName, downloadPicked } = this.props

    const columns = downloadPicked ? this.pickedColumns() : this.columns()
    const headers = columns.map(c => c.name)
    const valueKeys = columns.map(c => c.dataKey)

    let csvContent = ''

    headers.forEach((h) => {
      csvContent += `"${String(h).replace(/"/g, '""')}",`
    })
    csvContent = csvContent.slice(0, -1)
    csvContent += '\r\n'

    data.forEach((d) => {
      valueKeys.forEach((x) => {
        csvContent += `"${String(d[x]).replace(/"/g, '""')}",`
      })
      csvContent = csvContent.slice(0, -1)
      csvContent += '\r\n'
    })

    const url = `data:text/csv;charset=utf-8,${encodeURIComponent(csvContent)}`
    const link = document.createElement('a')
    link.setAttribute('href', url)
    link.setAttribute('download', `${downloadName}.csv`)
    document.body.appendChild(link)

    link.click()
    link.remove()
  }

  columns = () => {
    const { children, columns, data} = this.props
    const emptyData = []
    if (!children && !columns) {
      const columns = []
      if (!data.length) {
        return emptyData
      }
      Object.keys(data[0]).forEach(key => (key !== '_id' && columns.push({
        ...columnDefaultProps,
        name: key,
        dataKey: key,
        pickable: true,
        searchable: true,
        sortType: (Number.isInteger(data[0][key])) ? 'basic' : ((Number.isInteger(Date.parse(data[0][key]))) ? 'date' : 'string')
      })))
      return columns
    }

    if (Array.isArray(columns) && columns.length > 0) {
      // apply default here since columns are not DataTableColumn instances
      return columns.map(c => ({ ...columnDefaultProps, ...c }))
    }

    return (Array.isArray(children) ? children : [children])
      .filter(c => c.type === DataTableColumn || c.type.name === 'DataTableColumn')
      .map(c => c.props)
  }

  searchables = () => this.columns().filter(c => c.searchable).map(c => c.dataKey)

  pickables = () => this.columns().filter(c => c.pickable).map(c => c.name)

  pickedColumns = () => {
    const { picked } = this.state

    return this.columns().filter(c => !c.pickable || picked.includes(c.name))
  }

  onPageChange = (_, { activePage }) => {
    this.setState({ activePage })
  }

  handleSort = column => () => {
    const { sortColumn, sortDirection } = this.state
    if (column === sortColumn) {
      this.setState({
        activePage: 1,
        sortDirection:
          sortDirection === 'ascending' ? 'descending' : 'ascending',
      })
    } else {
      this.setState({
        activePage: 1,
        sortColumn: column,
        sortType: this.columns().find(col => col.dataKey === column).sortType,
        sortDirection: 'ascending',
      })
    }
  }

  handlePick = name => () => {
    const { picked } = this.state

    if (picked.includes(name)) {
      this.setState({ picked: picked.filter(c => c !== name) })
    } else {
      this.setState({ picked: [...picked, name] })
    }
  }

  getFilteredData() {
    const { data, search } = this.props
    const { searchInput } = this.state
    const text = searchInput.toLowerCase()
    const searchables = this.searchables()
    if (searchables.length === 0) {
      return data
    } else if (search) {
      return search(text, data, searchables)
    }
    return defaultSearch(text, data, searchables)
  }

  onSearchInputChange = (_, { value }) => {
    if (value && value.length > 2) {
      this.setState({ activePage: 1 })
    }
    this.setState({ searchInput: value.toLowerCase() })
  }

  createRowClickListener = (rowData) => (mouseEvent) => {
    return this.props.onRowClick(mouseEvent, rowData)
  }

  renderCell = (row, col) => {
    if (col.nullTemplate && (row[col.dataKey] === null || row[col.dataKey] === undefined)) {
      return col.nullTemplate
    }
    if (col.format) {
      return numeral(row[col.dataKey]).format(col.format)
    }
    if (col.template) {
      return col.template(row[col.dataKey], row)
    }

    return row[col.dataKey]
  }

  render() {
    const {
      data,
      download,
      perPage,
      onRowClick,
      isRowActive,
      emptySearchMsg,
      noColumnsMsg,
    } = this.props
    const tableProps = Object.entries(this.props)
      .filter(([key]) => !Object.keys(propTypes).includes(key))
      .reduce((acc, [key, value]) => {
        acc[key] = value
        return acc
      }, {})
    const { activePage, sortColumn, sortDirection, sortType, searchInput, picked } = this.state

    // set unique row key
    let id = 0
    data.forEach((row) => {
      id += 1
      row._id = id
    })

    // searching
    const searchables = this.searchables()
    const filteredData = searchInput && searchInput.length > 2
      ? this.getFilteredData()
      : data

    const columns = this.pickedColumns()

    const sortedData = sortType ? filteredData.sort(
      (a, b) => customSort(
        sortType,
        sortDirection)(a[sortColumn], b[sortColumn])
    ) : filteredData

    // pagination
    const offset = perPage * activePage
    const totalPages = Math.ceil(sortedData.length / perPage)
    const paginatedData = sortedData.filter((d, i) => i >= (offset - perPage) && i < offset)

    // pick/toggle
    const pickables = this.pickables()

    return (
      <div style={{ marginTop: '1em', paddingBottom: '1em' }}>
        <div style={{ width: '100%', overflowX: 'auto' }}>
          {(download || searchables.length > 0) && (
            <div style={{
              padding: '8px',
              border: '1px solid rgba(34, 36, 38, 0.15)',
              borderBottom: 0,
              borderRadius: '4px 4px 0px 0px',
              background: '#F9FAFB',
              overflow: 'auto',
            }}
            >
              <Form size='mini'>
                {searchables.length > 0 && (
                  <Form.Input
                    type='text'
                    placeholder='Search...'
                    value={searchInput}
                    onChange={this.onSearchInputChange}
                    size='medium'
                    icon='search'
                  />
                )}
                <div>
                  {pickables.length > 0 && (
                    <Button.Group size='mini'>
                      {pickables.map(name => (
                        <Button
                          key={name}
                          content={name}
                          primary={picked.includes(name)}
                          onClick={this.handlePick(name)}
                        />
                      ))}
                    </Button.Group>
                  )}
                  {download && (
                    <Button
                      floated='right'
                      size='mini'
                      onClick={this.downloadReport}
                      color='blue'
                      icon='download'
                    />
                  )}
                </div>
              </Form>
            </div>
          )}
          <Table
            sortable
            selectable
            style={{ marginTop: 0, borderRadius: 0 }}
            {...tableProps}
          >
            <Table.Header>
              <Table.Row>
                {columns.map(col => (
                  <Table.HeaderCell
                    key={col.dataKey || col.name}
                    onClick={col.sortable ? this.handleSort(col.dataKey) : null}
                    sorted={sortColumn === col.dataKey ? sortDirection : null}
                  >
                    {col.name}
                  </Table.HeaderCell>
                ))}
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {
                paginatedData.length === 0 &&
                columns.length>0 &&
                <Table.Row textAlign='center'>
                  <Table.HeaderCell colSpan={columns.length}>{emptySearchMsg}</Table.HeaderCell>
                </Table.Row>
              }
              {
                columns.length === 0 &&
                <Table.Row textAlign='center'>
                  <Table.HeaderCell colSpan={columns.length}>{noColumnsMsg}</Table.HeaderCell>
                </Table.Row>
              }
              {paginatedData.map(row => (
                <Table.Row
                  key={row._id}
                  active={typeof isRowActive === 'function' ? isRowActive(row) : undefined}
                  onClick={typeof onRowClick === 'function' ? this.createRowClickListener(row) : undefined}>
                  {columns.map((col) => {
                    // split out generic ...celProps passed-through similar to ...tableProps
                    const cellProps = { ...col }
                    const colProps = {}
                    colPropKeys.forEach((key) => {
                      if (key in cellProps) {
                        colProps[key] = cellProps[key]
                        delete cellProps[key]
                      }
                    })
                    return (
                      <Table.Cell key={col.dataKey || col.name} {...cellProps}>
                        {this.renderCell(row, colProps)}
                      </Table.Cell>
                    )
                  })}
                </Table.Row>
              ))}
            </Table.Body>
          </Table>
        </div>
        {totalPages > 1 && (
          <div style={{
            padding: '8px',
            border: '1px solid rgba(34, 36, 38, 0.15)',
            borderTop: 0,
            borderRadius: '0px 0px 4px 4px',
            background: '#F9FAFB',
          }}
          >
            <Pagination
              activePage={activePage}
              totalPages={totalPages}
              onPageChange={this.onPageChange}
            />
          </div>
        )}
      </div>
    )
  }
}

DataTable.Column = DataTableColumn

DataTable.propTypes = propTypes
DataTable.defaultProps = defaultProps

export default DataTable
