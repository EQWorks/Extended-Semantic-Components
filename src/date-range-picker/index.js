import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { momentObj } from 'react-moment-proptypes'

import { Button, Popup } from 'semantic-ui-react'
// TODO: take this out and make it a standalone open-source NPM package
import { DayPickerRangeController } from 'react-dates'
import moment from 'moment'


const propTypes = {
  endDate: momentObj,
  startDate: momentObj,
  onSelection: PropTypes.func.isRequired,
  btnProps: PropTypes.object,
  position: PropTypes.string,
  maximumNights: PropTypes.number,
  blockedStart: momentObj,
  blockedEnd: momentObj,
}

const defaultProps = {
  endDate: null,
  startDate: null,
  btnProps: {},
  position: 'bottom center',
  maximumNights: 0,
  blockedStart: null,
  blockedEnd: null,
}

class DateRangePicker extends Component {
  constructor(props) {
    super(props)
    const { startDate, endDate } = props
    this.state = {
      focusedInput: null,
      isOpen: false,
      startDate: startDate ? moment(startDate) : startDate,
      endDate: endDate ? moment(endDate) : endDate,
    }
  }

  componentDidUpdate(prevProps) {
    const { startDate, endDate } = this.props
    const { startDate: _start, endDate: _end } = prevProps
    // update local versions only when they're changed
    if ((startDate || _start) && !moment(_start).isSame(startDate)) {
      this.setState({ startDate })
    }
    if ((endDate || _end) && !moment(_end).isSame(endDate)) {
      this.setState({ endDate })
    }
  }

  open = (e) => {
    e.preventDefault()
    this.setState({
      isOpen: true,
      startDate: null,
      endDate: null,
    })
  }

  close = () => {
    const { startDate, endDate } = this.props
    this.setState({
      isOpen: false,
      startDate: startDate ? moment(startDate) : null,
      endDate: endDate ? moment(endDate) : null,
    })
  }

  onOpen = () => {
    this.setState({ focusedInput: 'startDate' })
  }

  onFocusChange = (focusedInput) => {
    if (!focusedInput) {
      this.setState({ focusedInput: 'startDate' })
    } else {
      this.setState({ focusedInput })
    }
  }

  onDatesChange = ({ startDate, endDate }) => {
    const endStart = moment(endDate).startOf('day')
    const currEndStart = moment(this.state.endDate).startOf('day')
    if (!(endDate === null || endStart.isSame(currEndStart))) {
      this.close()
      this.props.onSelection({ startDate, endDate })
    }
    this.setState({ startDate, endDate })
  }

  isOutsideRange = (day) => {
    const { focusedInput, startDate } = this.state
    const { maximumNights, blockedStart, blockedEnd } = this.props

    const dayStart = moment(day).startOf('day')
    const startStart = moment(startDate).startOf('day')

    let blocked = false

    if (maximumNights > 0 && moment(startDate).isValid() && focusedInput === 'endDate') {
      const maxLimit = moment(startStart).add(maximumNights, 'days')
      blocked = blocked || dayStart.isSameOrAfter(maxLimit)
    }

    if (moment(blockedStart).isValid()) {
      blocked = blocked || dayStart.isBefore(moment(blockedStart).startOf('day'))
    }

    if (moment(blockedEnd).isValid()) {
      blocked = blocked || dayStart.isAfter(moment(blockedEnd).startOf('day'))
    }

    return blocked
  }

  renderTrigger = () => {
    const { btnProps } = this.props
    const { startDate, endDate } = this.state
    const endDateString = endDate ? endDate.format('ll') : 'Select End Date'

    const durationString = startDate ?
      `${(startDate.format('ll'))} \u2192 ${endDateString}` :
      'Select Date Range'

    return (
      <Button
        basic
        {...btnProps}
        content={durationString}
        onClick={this.open}
      />
    )
  }

  renderPicker = () => {
    const { startDate, endDate, focusedInput } = this.state

    return (
      <DayPickerRangeController
        startDate={startDate}
        endDate={endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={focusedInput}
        onFocusChange={this.onFocusChange}
        isOutsideRange={this.isOutsideRange}
        numberOfMonths={2}
        hideKeyboardShortcutsPanel
        minimumNights={0}
      />
    )
  }

  render() {
    const { position } = this.props

    return (
      <Popup
        basic
        trigger={this.renderTrigger()}
        content={this.renderPicker()}
        on='click'
        onOpen={this.onOpen}
        onClose={this.close}
        open={this.state.isOpen}
        position={position}
        style={{ maxWidth: '800px' }}
      />
    )
  }
}

DateRangePicker.propTypes = propTypes
DateRangePicker.defaultProps = defaultProps

export default DateRangePicker
