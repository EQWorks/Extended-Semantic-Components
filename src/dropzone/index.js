import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import { useDropzone } from 'react-dropzone'
import styled from 'styled-components'


const propTypes = {
  /* core */
  /** File upload handler */
  onDrop: PropTypes.func.isRequired,
  /**
  A comma-separated list of unique content type specifiers,
  For example: .jpg, .png, .doc, .csv, .pdf, image/*, video/*, audio/*
  See more in https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#Unique_file_type_specifiers
  */
  accept: PropTypes.string,

  /* file preview */
  /** renderPreview(acceptedFiles) */
  renderPreview: PropTypes.func,
  /** if true, use renderPreview for init rendering */
  initWithPreview: PropTypes.bool,

  /* custom message */
  defaultMsg: PropTypes.string,
  activeMsg: PropTypes.string,
  rejectMsg: PropTypes.string,
  note: PropTypes.string,
}

const defaultProps = {
  renderPreview: null,
  accept: null,
  initWithPreview: false,
  defaultMsg: 'Drop File Here or Click to Upload',
  activeMsg: 'This File Is Authorized',
  rejectMsg: 'This File Is Not Authorized',
  note: '',
}

const Dropzone = ({
  onDrop: customOnDrop,
  accept,
  renderPreview, initWithPreview,
  defaultMsg, activeMsg, rejectMsg, note,
}) => {
  const onDrop = useCallback(customOnDrop, [])

  const {
    getRootProps,
    getInputProps,
    isDragActive,
    isDragAccept,
    isDragReject,
    acceptedFiles,
  } = useDropzone({ onDrop, accept })

  const renderChildren = () => {
    if (initWithPreview || acceptedFiles.length) {
      return renderPreview ? renderPreview(acceptedFiles) : (
        acceptedFiles.map((f, i) => <div key={`${i}_${f.name}`}>{f.name}</div>)
      )
    }

    let desc
    if (isDragReject) {
      desc = rejectMsg
    } else if (isDragActive) {
      desc = activeMsg
    } else {
      desc = defaultMsg
    }

    return (
      <DropzoneDescDiv>
        {desc}
        <DropzoneNoteDiv>
          {note}
        </DropzoneNoteDiv>
      </DropzoneDescDiv>
    )
  }

  return (
    <StyledDropzone {...getRootProps({ isDragActive, isDragAccept, isDragReject })}>
      <input {...getInputProps()} />
      {renderChildren()}
    </StyledDropzone>
  )
}

/* Styles */
const getBorderColor = (props) => {
  if (props.isDragReject) {
    return 'rgb(204, 102, 102)'
  }
  if (props.isDragAccept || props.isDragActive) {
    return 'rgb(102, 204, 102)'
  }
  return 'rgb(102, 102, 102)'
}

const StyledDropzone = styled.div`
  width: 100%;
  padding: 3.5rem 3.5rem;
  border-width: 1px;
  border-color: ${props => getBorderColor(props)};
  border-style: ${props => props.isDragActive ? 'solid' : 'dashed'};
  background-color: ${props => props.isDragActive ? 'rgb(238, 238, 238)' : ''};
  border-radius: 5px;
`

const DropzoneDescDiv = styled.h1`
  margin: 40px 0px !important;
  font-size: 20px;
  text-align: center;
  color: #646C7F;
  font-weight: 800;
`

const DropzoneNoteDiv = styled.p`
  margin-top: 1.5rem !important;
  font-weight: 100;
  font-size: 16px;
`

Dropzone.propTypes = propTypes
Dropzone.defaultProps = defaultProps

export default Dropzone
