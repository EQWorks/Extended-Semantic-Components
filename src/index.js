export { default as DateRangePicker } from './date-range-picker'
export { default as DataTable } from './data-table'
