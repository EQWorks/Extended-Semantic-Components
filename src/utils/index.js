export { sort, getDefaultSortType } from './sort'
export { default as defaultSearch } from './search'
