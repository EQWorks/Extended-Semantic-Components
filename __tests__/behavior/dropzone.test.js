import React from 'react'
import renderer from 'react-test-renderer'

import { Dropzone } from '../../src'


test('Dropzone should render', () => {
  const component = renderer.create(
    <Dropzone onDrop={() => {}} />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Dropzone with custom message for CSV', () => {
  const component = renderer.create(
    <Dropzone
      onDrop={() => {}}
      accept='.csv, text/csv, application/vnd.ms-excel'
      defaultMsg='Drop your file here'
      activeMsg='I see you are trying to drop something'
      rejectMsg='Invalid file type'
      note='Only CSV files are allowed'
    />
  )

  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
