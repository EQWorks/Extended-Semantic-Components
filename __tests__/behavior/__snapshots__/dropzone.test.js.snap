// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropzone should render 1`] = `
<div
  className="sc-bdVaJa cpcuHa"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    accept={null}
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <h1
    className="sc-bwzfXH hzZAvg"
  >
    Drop File Here or Click to Upload
    <p
      className="sc-htpNat hRMbIi"
    >
      
    </p>
  </h1>
</div>
`;

exports[`Dropzone with custom message for CSV 1`] = `
<div
  className="sc-bdVaJa cpcuHa"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    accept=".csv, text/csv, application/vnd.ms-excel"
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <h1
    className="sc-bwzfXH hzZAvg"
  >
    Drop your file here
    <p
      className="sc-htpNat hRMbIi"
    >
      Only CSV files are allowed
    </p>
  </h1>
</div>
`;
